### ISSUE TITLE
[GPU Memory Issue] why does the memory go peak on the first training log?

### ISSUE THREAD

**Issue by nijkah** [open-mmlab/mmengine#651]
I am benchmarking the effectiveness of [`ZeroRedundancyOptimizer`](https://github.com/open-mmlab/mmengine/pull/551) which is merged recently.
But I found that I cannot utilize the memory reduction using `ZeroRedundancyOptimizer` *to increase the batch size*.

Using the config [`convnext-xlarge_upernet_8xb2-amp-160k_ade20k-640x640.py`](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/configs/convnext/convnext-xlarge_upernet_8xb2-amp-160k_ade20k-640x640.py),

I could get these results,

### batch size 2 x (4 A100 (40GB) GPUs)
|  Method  |      Time       | Second log Memory  |
| :------: | :-------------: | :------: |
| w/o AMP | 0.6614 | 18048|
| ZeRO w/o  AMP  | 0.6500 | 15808 |
| w/  AMP  | 0.3600 | 14854 |
| ZeRO w/  AMP  | 0.3500 | 12616 |

### Maximum batch size x (4 A100 (40GB) GPUs)
|  Method  |  Batch Size |    First log Memory       | Second log  Memory  |
| :------:  | :------: | :-------------: | :------: |
| w/  AMP  | 7 | 35741 |  31422 |
| w/  AMP  | 8 | 37866 |  34787 |
| ZeRO w/  AMP |8  | 37866  | 32528 |

The reduced amount of memory is close to the expected `(1.5GB * 1024) * (1+1) * (3/4) = 2304` for batch_size=4 and batch_size=8.

If the memory usage on the second iter log is correct, then I may be able to increase the batch size to 1 ~ 2 when using `ZeroRedundancyOptimizer`.
But I couldn't.

I wonder if this is expected behavior or abnormal behavior with a memory leak.

**Comment by HAOCHENYE** [comment]
We may invite @C1rN09  to have a look.

**Comment by C1rN09** [comment]
I checked my previous `UpperNet` training log and find the similar issue. I'm not sure if it's a bug in **MMEngine** or **MMSegmentation** or **ZeroRedundancyOptimizer**. I'll investigate this.

To make it more clear:
- Does the abnormal peak of memory consumption occur if you are using the default configuration (i.e. without ZeroRedundancyOptimizer and other modifications)?
- Could you try a model from other repositories (e.g. `mmdetection`, `mmclassification` etc.) with/without ZeroRedundancyOptimizer on your machine?

**Comment by nijkah** [comment]
> I checked my previous `UpperNet` training log and find the similar issue. I'm not sure if it's a bug in **MMEngine** or **MMSegmentation** or **ZeroRedundancyOptimizer**. I'll investigate this.
> 
> To make it more clear:
> 
> * Does the abnormal peak of memory consumption occur if you are using the default configuration (i.e. without ZeroRedundancyOptimizer and other modifications)?
> * Does it happen only with `AMP` enabled?
> * Could you try a model from other repositories (e.g. `mmdetection`, `mmclassification` etc.) with/without ZeroRedundancyOptimizer on your machine?

I'm still trying to identify the cases when the memory peak occurs.

1. Yes, it does. For example, training [upernet_r101_4xb4-160k_ade20k-512x512.py](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/configs/upernet/upernet_r101_4xb4-160k_ade20k-512x512.py) with 4 GPUs shows below log.
```10/31 08:56:47 - mmengine - INFO - Checkpoints will be saved to /data/workspace/hakjinlee/R-BIG22A/mmsegmentation/work_dirs/upernet_r101_4xb4-160k_ade20k-512x512 by HardDiskBackend.
10/31 08:57:43 - mmengine - INFO - Iter(train) [50/160000]  lr: 9.9973e-03  eta: 2 days, 1:37:19  time: 0.3585  data_time: 0.0105  memory: 21451  decode.loss_ce: 2.2373  decode.acc_seg: 52.6685  aux.loss_ce: 1.2076  aux.acc_seg: 33.5014  loss: 3.8896
10/31 08:58:01 - mmengine - INFO - Iter(train) [100/160000]  lr: 9.9945e-03  eta: 1 day, 8:51:18  time: 0.3617  data_time: 0.0110  memory: 8163  decode.loss_ce: 2.7752  decode.acc_seg: 41.1756  aux.loss_ce: 1.3059  aux.acc_seg: 32.8448  loss: 3.0869
10/31 08:58:19 - mmengine - INFO - Iter(train) [150/160000]  lr: 9.9917e-03  eta: 1 day, 3:07:26  time: 0.3583  data_time: 0.0110  memory: 8163  decode.loss_ce: 1.3035  decode.acc_seg: 63.9396  aux.loss_ce: 0.8193  aux.acc_seg: 40.4417  loss: 2.6328
```
2. Like above, it happens without AMP.
3. In the case of `mmdetection`, [faster-rcnn_r50_fpn_1x_coco.py](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/faster_rcnn/faster-rcnn_r50_fpn_1x_coco.py), [faster-rcnn_r101_fpn_1x_coco.py](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/faster_rcnn/faster-rcnn_r101_fpn_1x_coco.py), [configs/faster_rcnn/faster-rcnn_x101-64x4d_fpn_1x_coco.py](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/faster_rcnn/faster-rcnn_x101-64x4d_fpn_1x_coco.py) maintain the same memory usage, but [configs/swin/mask-rcnn_swin-s-p4-w7_fpn_amp-ms-crop-3x_coco.py](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/swin/mask-rcnn_swin-s-p4-w7_fpn_amp-ms-crop-3x_coco.py) shows peak.
```
10/31 09:54:19 - mmengine - INFO - Checkpoints will be saved to /data/workspace/hakjinlee/R-BIG22A/mmdetection/work_dirs/mask-rcnn_swin-s-p4-w7_fpn_amp-ms-crop-3x_coco by HardDiskBackend.
10/31 09:54:47 - mmengine - INFO - Epoch(train) [1][50/14659]  lr: 5.0000e-06  eta: 3 days, 9:50:34  time: 0.5584  data_time: 0.0392  memory: 6801  loss_rpn_cls: 0.6898  loss_rpn_bbox: 0.1007  loss_cls: 3.0107  acc: 96.8750  loss_bbox: 0.0333  loss_mask: 0.9933  loss: 4.8279
10/31 09:55:06 - mmengine - INFO - Epoch(train) [1][100/14659]  lr: 1.0000e-05  eta: 2 days, 19:57:41  time: 0.3690  data_time: 0.0276  memory: 6381  loss_rpn_cls: 0.5200  loss_rpn_bbox: 0.0832  loss_cls: 0.4881  acc: 97.8516  loss_bbox: 0.1232  loss_mask: 0.8137  loss: 2.0283
```

And also [`configs/swin/mask-rcnn_swin-t-p4-w7_fpn_1x_coco.py`](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/swin/mask-rcnn_swin-t-p4-w7_fpn_1x_coco.py) shows the memory peak.
```
10/31 09:59:00 - mmengine - INFO - Epoch(train) [1][50/14659]  lr: 5.0000e-06  eta: 1 day, 6:31:26  time: 0.6249  data_time: 0.0653  memory: 6758  loss_rpn_cls: 0.6517  loss_rpn_bbox: 0.1385  loss_cls: 3.5427  acc: 93.7500  loss_bbox: 0.0462  loss_mask: 0.9177  loss: 5.2968
10/31 09:59:17 - mmengine - INFO - Epoch(train) [1][100/14659]  lr: 1.0000e-05  eta: 23:27:40  time: 0.3360  data_time: 0.0363  memory: 6641  loss_rpn_cls: 0.4860  loss_rpn_bbox: 0.1169  loss_cls: 0.5998  acc: 97.1680  loss_bbox: 0.1522  loss_mask: 0.7871  loss: 2.1421
```

I am not sure about this :(

**Comment by nijkah** [comment]
`mmclassification` results.

`vit-base-p16_pt-64xb64_in1k-224.py`
```
10/31 10:44:14 - mmengine - INFO - Epoch(train) [1][10/5005]  lr: 4.7980e-07  eta: 86 days, 8:23:15  time: 4.9688  data_time: 2.9146  memory: 11225  grad_norm: 1.8124  loss: 6.9077
10/31 10:44:57 - mmengine - INFO - Epoch(train) [1][20/5005]  lr: 6.7958e-07  eta: 80 days, 12:56:42  time: 4.3002  data_time: 3.1700  memory: 11225  grad_norm: 1.9184  loss: 6.9078
```
`swin-tiny_16xb64_in1k.py`
```
10/31 10:47:34 - mmengine - INFO - Epoch(train) [1][10/5005]  lr: 1.0898e-06  eta: 78 days, 12:05:59  time: 4.5174  data_time: 2.5596  memory: 7154  grad_norm: 2.4795  loss: 6.9608
10/31 10:48:20 - mmengine - INFO - Epoch(train) [1][20/5005]  lr: 1.1896e-06  eta: 79 days, 21:44:10  time: 4.6787  data_time: 4.0350  memory: 7154  grad_norm: 2.4878  loss: 6.9830
```
resnext101-32x4d_8xb32_in1k.py
```
10/31 10:53:54 - mmengine - INFO - Epoch(train) [1][10/10010]  lr: 1.0000e-01  eta: 26 days, 3:20:44  time: 2.2562  data_time: 1.1350  memory: 5839  loss: 7.0360
10/31 10:54:14 - mmengine - INFO - Epoch(train) [1][20/10010]  lr: 1.0000e-01  eta: 24 days, 13:29:51  time: 1.9840  data_time: 1.6781  memory: 5839  loss: 7.2258
```

When I think the memory usage can vary in `mmdetection`, the reason for the memory peak can be based on `mmsegmentation` (not sure yet)

**Comment by C1rN09** [comment]
I found a [post](https://discuss.pytorch.org/t/larger-gpu-memory-usage-during-the-first-batch/58768) which might be relevant. It indicates that high memory consumption at first batch is caused by cudnn benchmarking. Seems like this is set to `False` in `mmcls` and `mmdet`, while `True` in `mmseg`. You may find this configuration in `default_runtime.py`. If this is the case, changing `cudnn_benchmark=False` may resolve this issue.

However, the post indicates that it won't result in OOM error. Have you encountered OOM error when trying to increase batch size?

**Comment by nijkah** [comment]
@C1rN09, it affects the maximum number of batch sizes when using `ZeroRedundancyOptimizer`. WOW 😮 !
With the config [convnext-xlarge_upernet_8xb2-amp-160k_ade20k-640x640.py](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/configs/convnext/convnext-xlarge_upernet_8xb2-amp-160k_ade20k-640x640.py) which I mentioned in the issue description,
the run with `batch_size=9` and `cudnn_benchmark=True` raises OOM when using A100 GPU`(40537MiB)`.

On the other hand, the run with `batch_size=9` and `cudnn_benchmark=False` can start the training well with the log,
```
11/01 01:01:34 - mmengine - INFO - Iter(train) [10/160000]  lr: 1.2211e-07  eta: 5 days, 8:01:49  time: 2.8809  data_time: 0.0206  memory: 35126  decode.loss_ce: 4.2248  decode.acc_seg: 0.1940  aux.loss_ce: 1.6856  aux.acc_seg: 1.2063  loss: 5.7015
11/01 01:01:46 - mmengine - INFO - Iter(train) [20/160000]  lr: 2.5777e-07  eta: 3 days, 19:21:15  time: 1.2306  data_time: 0.0186  memory: 35126  decode.loss_ce: 3.7377  decode.acc_seg: 0.3961  aux.loss_ce: 1.4882  aux.acc_seg: 2.0518  loss: 5.5250
11/01 01:01:58 - mmengine - INFO - Iter(train) [30/160000]  lr: 3.9342e-07  eta: 3 days, 7:18:31  time: 1.2429  data_time: 0.0228  memory: 35125  decode.loss_ce: 4.0037  decode.acc_seg: 0.2481  aux.loss_ce: 1.5900  aux.acc_seg: 1.2825  loss: 5.8229
```
Unfortunately, `batch_size=10` and `cudnn_benchmark=False` raises the OOM too.

`ZeroRedundancyOptimizer` seems not so effective when using it with `ConvNets` since the most of memory usage comes from the activations compared to the `NLP` or `MLP` models. 
Anyway, your information was so helpful. Thanks 😄

**Comment by wujiang0156** [comment]
I also have the same bug, when i trianed with MMSegmentation .

**Comment by Y-T-G** [comment]
This issue happens right after the last epoch for me. You can see the high memory usage in `val` iteration. And if you disable `val`, you still see it in the first iteration of the next epoch in `train`. But you don't see it between `val` and `train` for next epoch. So there's something in the last iteration of the `train` that causes the memory usage jump.


```
569 10/23 20:14:31 - mmengine - INFO - Epoch(val)  [4][  50/5000]    eta: 0:04:34  time: 0.0554  data_time: 0.0049  memory: 22951
570 10/23 20:14:34 - mmengine - INFO - Epoch(val)  [4][ 100/5000]    eta: 0:04:42  time: 0.0598  data_time: 0.0059  memory: 784
571 10/23 20:14:37 - mmengine - INFO - Epoch(val)  [4][ 150/5000]    eta: 0:04:39  time: 0.0576  data_time: 0.0080  memory: 788
```
